File: Data Structures & Algorithms/interleaving-string/submission-2.cpp
class Solution {
public:
    bool interleave(
        string &s1,
        string &s2,
        string &s3,
        int i,
        int j,
        vector<vector<int>> &dp
    ){
        int n1 = s1.size();
        int n2 = s2.size();
        int n3 = s3.size();
        if(i+j==n3){
            return true;
        }
        bool ans = false;
        if(i<n1 && j<n2 && dp[i][j]!=-1){
            return (bool)dp[i][j];
        }
        if(i<n1 && s1[i]==s3[i+j]){
            ans = ans || interleave(s1,s2,s3,i+1,j,dp);
        }
        if(j<n2 && s2[j]==s3[i+j]){
            ans = ans || interleave(s1,s2,s3,i,j+1,dp);
        }
        
        return dp[i][j]=(int)ans;
    }
    bool isInterleave(string s1, string s2, string s3) {
       int n1 = s1.size();
       int n2 = s2.size();
       int n3 = s3.size();

       if(n1+n2!=n3){
        return false;
       }
       vector<vector<int>> dp(n1+1,vector<int>(n2+1,-1));

       return interleave(s1,s2,s3,0,0,dp);
       
    }
};

File: Data Structures & Algorithms/4sum/submission-2.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        ans: List[List[int]] = []

        for i in range(0, len(nums)-3):
            if i > 0 and nums[i - 1] == nums[i]:
                continue
            for j in range(i+1, len(nums)-2):
                if j > i+1 and nums[j - 1] == nums[j]:
                    continue
                currtarget:int = target - (nums[i]+nums[j])
                start: int = j + 1
                end: int = len(nums) - 1

                while start < end:
                    sum: int = nums[start] + nums[end]
                    
                    if sum == currtarget:
                        ans.append([nums[i], nums[j], nums[start], nums[end]])
                        start += 1
                        end -= 1
                        while start<end and start>0 and nums[start]==nums[start-1]:
                            start+=1
                        while start<end and end<len(nums)-1 and nums[end]==nums[end+1]:
                            end-=1
                    elif sum < currtarget:
                        start += 1
                    else:
                        end -= 1
        return ans
